File: go/delete-node-in-bst.go
package leetcode

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func deleteNode(root *TreeNode, key int) *TreeNode {
	if root == nil {
		return nil
	}

	if root.Val < key {
		root.Right = deleteNode(root.Right, key)
	} else if root.Val > key {
		root.Left = deleteNode(root.Left, key)
	} else {
		if root.Left == nil && root.Right == nil { // has no children
			return nil
		} else if root.Left == nil && root.Right != nil { // has only right child
			return root.Right
		} else if root.Right == nil && root.Left != nil { // has only left child
			return root.Left
		} else { // has both left and right children
			curr := root.Right
			for curr.Left != nil {
				curr = curr.Left
			}
			root.Val = curr.Val
			root.Right = deleteNode(root.Right, curr.Val)
		}
	}
	return root
}

File: python/delete-node-in-bst.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root: return None
        
        if root.val < key: root.right = self.deleteNode(root.right, key)
        elif root.val > key: root.left = self.deleteNode(root.left, key)
        
        else:
            if root.left == None and root.right == None: return None # doesn't have any children
            if root.left == None and root.right: return root.right # has only right chilren
            if root.right == None and root.left: return root.left # has only left children

            # below is the case when it has both left and right children
            curr = root.right
            while curr.left:
                curr = curr.left

            root.val = curr.val
            root.right = self.deleteNode(root.right, curr.val)
        return root
        